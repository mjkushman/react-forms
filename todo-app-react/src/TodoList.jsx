import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import { useState } from "react";
import {v4 as uuid} from 'uuid';



const TodoList = () => {
    // define initial state
    const INITIAL_STATE = [
        {task:"Create a list of todos", id:uuid()}
    ]

    //instantiate state for todo list
    const [todos, setTodos] = useState(INITIAL_STATE)

    // function to submit the todo form
    const addTodo = (formData) => {
        setTodos((todos) => ([...todos, {...formData, id:uuid() }]
        ))
    }

    // function to handle removing todo from state
    const removeTodo = (todo) => {
        setTodos(todos.filter(t => t !== todo))
    }


    //return the component
    return (
        <>
        <h2>To-Do List</h2>
        <NewTodoForm addTodo={addTodo}/>
        <ul>
        {todos.map((todo) => <Todo task={todo.task} key={todo.id} removeTodo={ ()=> removeTodo(todo) }/>)}
        </ul>
        </>
    )
}

export default TodoList