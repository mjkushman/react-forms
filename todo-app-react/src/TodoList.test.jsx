import React from 'react'
import { render, fireEvent } from '@testing-library/react';
import {describe, test, expect} from 'vitest';
import TodoList from './TodoList.jsx'


// Smoke test
test('TodoList Renders without going up in smoke', ()=> {
  render(<TodoList />)
})

// snapshot test
test("snapshot match", ()=> {
  const {asFragment} = render(<TodoList/>);

  expect(asFragment()).toMatchSnapshot();
})


test("Form should add and delete items", ()=> {
    // renders the list component, which includes a form
    const {queryByText, getByLabelText} = render(<TodoList />);

    const input = getByLabelText("Task"); // selects the input field
    const button = queryByText("Add"); // selects the add button

    // make sure my task isn't already in the dom
    expect(queryByText('Pass this test')).toBeNull();

    // mock adding a task
    fireEvent.change(input,{target:{value:"Pass this test"} } )
    fireEvent.click(button)

    // check for new task
    expect(queryByText('Pass this test')).not.toBeNull()
})

test("Form should delete items", ()=> {
    // renders the list component, which includes a form
    const {queryByText, getByText} = render(<TodoList />);

    const todo = getByText("Create a list of todos"); // selects the input field
    const button = queryByText("X"); // selects the add button

    expect(todo).toBeTruthy()

    // mock adding a task
    fireEvent.click(button)

    // check for new task
    expect(queryByText('Pass this test')).not.toBeTruthy()
})