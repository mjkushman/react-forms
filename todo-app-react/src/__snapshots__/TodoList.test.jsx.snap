// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot match 1`] = `
<DocumentFragment>
  <h2>
    To-Do List
  </h2>
  <form>
    <label
      for="task"
    >
       Task 
    </label>
    <input
      id="task"
      name="task"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </form>
  <ul>
    <li>
      Create a list of todos
      <button>
        X
      </button>
    </li>
  </ul>
</DocumentFragment>
`;
